File: s01/s01_22/src/main.exs
defmodule Bank do
  defstruct users: %{}

  def start() do
    %Bank{}
  end

  def register(bank, username, password) do
    if Map.has_key?(bank.users, username) do
      IO.puts("用户名已存在！")
      bank
    else
      new_user = %{password: password, balance: 0.0}
      updated_users = Map.put(bank.users, username, new_user)
      IO.puts("用户 #{username} 注册成功！")
      %Bank{bank | users: updated_users}
    end
  end

  def login(bank, username, password) do
    case Map.get(bank.users, username) do
      nil ->
        IO.puts("用户不存在！")
        {:error, bank}

      user ->
        if user.password == password do
          IO.puts("登录成功！")
          {:ok, bank, username}
        else
          IO.puts("密码错误！")
          {:error, bank}
        end
    end
  end

  def deposit(bank, username, amount) do
    user = Map.get(bank.users, username)
    updated_user = %{user | balance: user.balance + amount}
    updated_users = Map.put(bank.users, username, updated_user)
    IO.puts("入金成功！当前余额：#{updated_user.balance}")
    %Bank{bank | users: updated_users}
  end

  def transfer(bank, from, to, amount) do
    if Map.has_key?(bank.users, to) do
      from_user = Map.get(bank.users, from)
      to_user = Map.get(bank.users, to)

      if from_user.balance < amount do
        IO.puts("余额不足！")
        bank
      else
        updated_from_user = %{from_user | balance: from_user.balance - amount}
        updated_to_user = %{to_user | balance: to_user.balance + amount}

        updated_users =
          bank.users
          |> Map.put(from, updated_from_user)
          |> Map.put(to, updated_to_user)

        IO.puts("转账成功！#{from} 向 #{to} 转账 #{amount} 元")
        %Bank{bank | users: updated_users}
      end
    else
      IO.puts("接收用户不存在！")
      bank
    end
  end

  def show_all_users(bank) do
    Enum.each(bank.users, fn {username, user} ->
      IO.puts("用户: #{username}, 余额: #{user.balance}")
    end)
  end
end

defmodule Main do
  def start() do
    bank = Bank.start()
    loop(bank, nil)
  end

  defp loop(bank, nil) do
    IO.puts("请选择操作: 1. 注册 2. 登录 3. 查询所有用户 4. 退出")
    choice = IO.gets("") |> String.trim() |> String.to_integer()

    case choice do
      1 ->
        {username, password} = get_credentials()
        new_bank = Bank.register(bank, username, password)
        loop(new_bank, nil)

      2 ->
        {username, password} = get_credentials()

        case Bank.login(bank, username, password) do
          {:ok, new_bank, username} ->
            loop(new_bank, username)

          {:error, new_bank} ->
            loop(new_bank, nil)
        end

      3 ->
        Bank.show_all_users(bank)
        loop(bank, nil)

      4 ->
        IO.puts("退出系统")
        :ok

      _ ->
        IO.puts("无效操作，请重新选择")
        loop(bank, nil)
    end
  end

  defp loop(bank, logged_in_user) do
    IO.puts("请选择操作: 1. 入金 2. 转账 3. 注销 4. 退出")
    choice = IO.gets("") |> String.trim() |> String.to_integer()

    case choice do
      1 ->
        IO.write("请输入入金额: ")
        amount = IO.gets("") |> String.trim() |> String.to_float()
        new_bank = Bank.deposit(bank, logged_in_user, amount)
        loop(new_bank, logged_in_user)

      2 ->
        IO.write("请输入接收用户名: ")
        to = IO.gets("") |> String.trim()
        IO.write("请输入转账金额: ")
        amount = IO.gets("") |> String.trim() |> String.to_float()
        new_bank = Bank.transfer(bank, logged_in_user, to, amount)
        loop(new_bank, logged_in_user)

      3 ->
        IO.puts("已注销 #{logged_in_user} 的登录。")
        loop(bank, nil)

      4 ->
        IO.puts("退出系统")
        :ok

      _ ->
        IO.puts("无效操作，请重新选择")
        loop(bank, logged_in_user)
    end
  end

  defp get_credentials() do
    IO.write("请输入用户名: ")
    username = IO.gets("") |> String.trim()
    IO.write("请输入密码: ")
    password = IO.gets("") |> String.trim()
    {username, password}
  end
end

Main.start()

File: s01/s01_22/src/main.php
<?php

class User {
    public $password;
    public $balance;

    public function __construct($password, $balance = 0) {
        $this->password = $password;
        $this->balance = $balance;
    }
}

class Bank {
    private $users = [];

    public function register($username, $password) {
        if (isset($this->users[$username])) {
            echo "用户名已存在！\n";
        } else {
            $this->users[$username] = new User($password);
            echo "用户 {$username} 注册成功！\n";
        }
    }

    public function login($username, $password) {
        if (isset($this->users[$username])) {
            if ($this->users[$username]->password === $password) {
                echo "登录成功！\n";
                return true;
            } else {
                echo "密码错误！\n";
            }
        } else {
            echo "用户不存在！\n";
        }
        return false;
    }

    public function deposit($username, $amount) {
        if (isset($this->users[$username])) {
            $this->users[$username]->balance += $amount;
            echo "入金成功！当前余额：{$this->users[$username]->balance}\n";
        } else {
            echo "用户不存在！\n";
        }
    }

    public function transfer($from, $to, $amount) {
        if (!isset($this->users[$to])) {
            echo "接收用户不存在！\n";
            return;
        }

        if (isset($this->users[$from])) {
            if ($this->users[$from]->balance < $amount) {
                echo "余额不足！\n";
            } else {
                $this->users[$from]->balance -= $amount;
                $this->users[$to]->balance += $amount;
                echo "转账成功！{$from} 向 {$to} 转账 {$amount} 元\n";
            }
        } else {
            echo "发送用户不存在！\n";
        }
    }

    public function showAllUsers() {
        foreach ($this->users as $username => $user) {
            echo "用户: {$username}, 余额: {$user->balance}\n";
        }
    }
}

function getInput($prompt) {
    echo $prompt;
    return trim(fgets(STDIN));
}

$bank = new Bank();
$loggedInUser = null;

while (true) {
    if ($loggedInUser === null) {
        echo "请选择操作: 1. 注册 2. 登录 3. 查询所有用户 4. 退出\n";
        $choice = getInput("");

        switch ($choice) {
            case '1':
                $username = getInput("请输入用户名: ");
                $password = getInput("请输入密码: ");
                $bank->register($username, $password);
                break;
            case '2':
                $username = getInput("请输入用户名: ");
                $password = getInput("请输入密码: ");
                if ($bank->login($username, $password)) {
                    $loggedInUser = $username;
                }
                break;
            case '3':
                $bank->showAllUsers();
                break;
            case '4':
                exit("退出系统\n");
            default:
                echo "无效操作，请重新选择\n";
        }
    } else {
        echo "请选择操作: 1. 入金 2. 转账 3. 注销 4. 退出\n";
        $choice = getInput("");

        switch ($choice) {
            case '1':
                $amount = getInput("请输入入金额: ");
                $bank->deposit($loggedInUser, (float)$amount);
                break;
            case '2':
                $to = getInput("请输入接收用户名: ");
                $amount = getInput("请输入转账金额: ");
                $bank->transfer($loggedInUser, $to, (float)$amount);
                break;
            case '3':
                echo "已注销 {$loggedInUser} 的登录。\n";
                $loggedInUser = null;
                break;
            case '4':
                exit("退出系统\n");
            default:
                echo "无效操作，请重新选择\n";
        }
    }
}

